util/misc.py

```diff
@@ -410,4 +412,118 @@ def which(application_binary_name):
         full_path=os.path.join(path, application_binary_name)
         if os.path.exists(full_path) and os.access(full_path, os.X_OK):
             link_resolved_path = os.path.realpath(full_path)
-            return link_resolved_path
\ No newline at end of file
+            return link_resolved_path
+
+def is_nonstr_iterable(x):
+    '''Tests whether `x` is an Iterable other than a string'''
+    return isinstance(x, collections.Iterable) and not isinstance(x,str)
+
+def make_seq(x):
+    '''Return a tuple containing the items in `x`, or containing just `x` if `x` is a non-string iterable.  Convenient
+    for uniformly writing iterations over parameters that may be passed in as either an item or a tuple/list of items.
+    Note that if `x` is an iterator, it will be concretized.
+    '''
+    return tuple(x) if is_nonstr_iterable(x) else (x,)
+
+def flatten_dict(d):
+    '''Return a new dict `r`, where r[(k1,k2,...,kn)]==v iff d[k1][k2]...[kn]==v.'''
+    result = dict()
+    for k, v in d.items():
+        if isinstance(v, collections.Mapping):
+            for k_rest, v_actual in flatten_dict(v).items():
+                result[(k,)+k_rest] = v_actual
+        else:
+            result[(k,)] = v
+    return result
+
+def unflatten_dict(d):
+    '''Reverse operation to `flatten_dict`'''
+    result = dict()
+    for k, v in d.items():
+        # if k is (k1,k2,...kN), then ensure that result[k1][k2]...[kN]=v, creating intermediate dicts as needed
+        functools.reduce(lambda d,k: d.setdefault(k,dict()), k[:-1], result)[k[-1]]=v
+    return result
+
+def load_yaml_or_json(fname):
+    '''Load a dictionary from either a yaml or a json file'''
+    with open(fname) as f:
+        if fname.upper().endswith('.YAML'): return yaml.safe_load(f) or {}
+        if fname.upper().endswith('.JSON'): return json.load(f) or {}
+        raise TypeError('Unsupported dict file format: ' + fname)
+
+def load_config(cfg, include_directive='include', std_includes=[], param_renamings={}):
```

Comment:
Probably shouldn't set the default to a mutable type; better to set it to `None` and then in the scope of the function initialize it as empty:

```
if std_includes is None:
        std_includes = []
```